pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,352 @@
+	if itops.BatchSize == 0 {
+		itops.BatchSize = defaultBatchSize // Default value for batch sizes. Possibly needs more testing to find an optimal default
+	}
+	if itops.BatchSize > int(g.m.MaxEntries()) {
+		itops.BatchSize = int(g.m.MaxEntries())
+	}
+
+	if BatchAPISupported() && !g.isPerCPU() && !itops.ForceSingleItem {
+		it := &genericMapBatchIterator[K, V]{
+			m:                            g.m,
+			batchSize:                    itops.BatchSize,
+			keys:                         make([]K, itops.BatchSize),
+			values:                       make([]V, itops.BatchSize),
+			valueTypeCanUseUnsafePointer: g.valueTypeCanUseUnsafePointer(),
+		}
+
+		// Do an initial copy of the keys/values slices to avoid allocations
+		it.keysCopy = it.keys
+		it.valuesCopy = it.values
+
+		return it
+	}
+
+	return &genericMapItemIterator[K, V]{
+		it:                           g.m.Iterate(),
+		valueTypeCanUseUnsafePointer: g.valueTypeCanUseUnsafePointer(),
+	}
+}
+
+// genericMapItemIterator is an iterator for a map that returns a single item at a time
+type genericMapItemIterator[K interface{}, V interface{}] struct {
+	it                           *ebpf.MapIterator
+	valueTypeCanUseUnsafePointer bool
+}
+
+// Next fills K and V with the next key/value pair in the map. It returns false if there are no more elements
+func (g *genericMapItemIterator[K, V]) Next(key *K, value *V) bool {
+	// we resort to unsafe.Pointers because by doing so the underlying eBPF
+	// library avoids marshaling the key/value variables while traversing the map
+	// However, in some cases (slices, structs) we need to pass the variable directly
+	// so that the library detects the type correctly
+	if g.valueTypeCanUseUnsafePointer {
+		return g.it.Next(unsafe.Pointer(key), unsafe.Pointer(value))
+	}
+
+	return g.it.Next(unsafe.Pointer(key), value)
+}
+
+// Err returns the last error that happened during iteration. Should be checked
+// after completing the iteration, as it can report issues such as wrong types
+// being passed to the Next() method or aborted iterations, which would be perceived
+// as empty/partial map iterations.
+func (g *genericMapItemIterator[K, V]) Err() error {
+	return g.it.Err()
+}
+
+// genericMapBatchIterator is an iterator for a map that, under the hood, uses BatchLookup to reduce
+// the number of syscalls
+type genericMapBatchIterator[K interface{}, V interface{}] struct {
+	m                            *ebpf.Map        // Map to iterate
+	batchSize                    int              // Number of items to fetch per batch
+	cursor                       ebpf.BatchCursor // Cursor that maintains the state of the iteration
+	keys                         []K              // Buffer for storing the keys of the current batch
+	values                       []V              // Buffer for storing the values of the current batch
+	keysCopy                     any              // A pointer to keys of type "any", used to avoid allocations when calling BatchLookup
```

Comment:
I think we may be able to get rid of these, but I ran into a bug in the upstream library: https://github.com/cilium/ebpf/issues/1290